src/index.js

```diff
@@ -1,35 +1,329 @@
+                      const block = ws.getBlockById(id);
+                      if (block) {
+                        apply(block);
+                      }
+                    });
+                    Blockly.Events.setGroup(false);
+                    return;
+                  }
+                }
+                if (!ws.doubleClickPid_) {
+                  ws.doubleClickBlock_ = Blockly.selected.id;
+                  ws.doubleClickPid_ = setTimeout(function() {
+                    ws.doubleClickPid_ = undefined;
+                  }, 500);
+                }
+              }
+            }
+          };
+          wrappedFunc.isWrapped = true;
+          return wrappedFunc;
+        }
+      })(Blockly.Gesture.prototype.handleWsStart);
+    } else {
+      Blockly.Gesture.prototype.handleWsStart = origHandleWsStart;
+    }
+  }
+
+  /**
+   * Maintain the selected blocks set list when updating.
+   * @param {!Blockly.BlockSvg} block The block to update.
+   * @private
+   */
+  updateBlocks_(block) {
+    if (block &&
+      block.isDeletable() &&
+      block.isMovable()) {
+      if (blockSelection.has(block.id)) {
+        blockSelection.delete(block.id);
+        this.justDeletedBlock_ = block;
+        block.pathObject.updateSelected(false);
+      } else {
+        blockSelection.add(block.id);
+        this.justDeletedBlock_ = null;
+        block.pathObject.updateSelected(true);
+        block.bringToFront();
+      }
+      console.log(blockSelection);
+    }
+  }
+
+  /**
+   * Handle workspace events.
+   * @param {!Event} e Blockly event.
+   * @private
+   */
+  eventListener_(e) {
+    // on Block Selected
+    if (e.type === Blockly.Events.SELECTED) {
+      if (!inMultipleSelectionMode) {
+        if (!Blockly.selected ||
+          (Blockly.selected && !blockSelection.has(Blockly.selected.id))) {
+          // When not in multiple selection mode and Blockly selects a block not
+          // in currently selected set or unselects, clear the selected set.
+          blockSelection.forEach((id) => {
+            const element = this.workspace_.getBlockById(id);
+            if (element) {
+              element.pathObject.updateSelected(false);
+            }
+          });
+          blockSelection.clear();
+          this.updateBlocks_(Blockly.selected);
+        }
+      } else if (this.justDeletedBlock_ && Blockly.selected &&
+        Blockly.selected.id === this.justDeletedBlock_.id) {
+        // Update the Blockly selected block when that block is
+        // no longer selected in our set.
+        if (blockSelection.size) {
+          Blockly.common.setSelected(
+              this.workspace_.getBlockById(blockSelection.keys().next().value));
+        } else {
+          Blockly.common.setSelected(null);
+        }
+        this.justDeletedBlock_.pathObject.updateSelected(false);
+        this.justDeletedBlock_ = null;
+      }
+      // Update the selection highlight.
```

Comment:
Why do you need to update the selection highlight in bulk if you're updating individual blocks individual blocks as they're selected/unselected?

Reply:
Okay, just checked that only update the block that the blockly.selected unselects is enough

Reply:
We may still want to do this here, otherwise such as in the multiple selection shortcut copy and paste, it's not easy to keep all those selected. Maybe it can be optimized later, but currently we just keep the original as it's an easy way to avoid bugs and I can't see any performance issues.